### ISSUE TITLE
storage: TestStoreRangeDownReplicate fatal: snapshot message from Raft.Ready 7c54e4eb doesn't match outgoing snapshot UUID 00000000.

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#9603]
> F160928 18:50:24.140751 376 storage/replica.go:2361  [s1,r2:{"m"-/Max}] programming error: snapshot message from Raft.Ready 96ecfeaf doesn't match outgoing snapshot UUID d8a1496f.
> goroutine 376 [running]:

```
I160928 18:50:23.883979 7566 storage/replica_raftstorage.go:580  [s4] [n4,s4,r2:{"m"-/Max}]: with replicaID [?], applied preemptive snapshot in 0.002s
I160928 18:50:23.890931 54 storage/replica_command.go:3278  change replicas: read existing descriptor range_id:2 start_key:"m" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:2 store_id:2 replica_id:2 > replicas:<node_id:3 store_id:3 replica_id:3 > next_replica_id:4
W160928 18:50:23.977231 8546 storage/stores.go:214  range not contained in one range: [/Meta2/Max,"m\x00"), but have [/Min,"m")
I160928 18:50:24.013904 9008 storage/replica.go:1690  [s1,r2:{"m"-/Max}] proposing ADD_REPLICA {NodeID:4 StoreID:4 ReplicaID:4} for range 2: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2} {NodeID:3 StoreID:3 ReplicaID:3} {NodeID:4 StoreID:4 ReplicaID:4}]
I160928 18:50:24.103015 9780 storage/replica_raftstorage.go:447  generated snapshot 96ecfeaf for range 2 at index 14 in 256.568µs.
W160928 18:50:24.129116 376 storage/replica.go:1922  [s1,r2:{"m"-/Max}] Abandoning unsent snapshot {96ecfeaf-e0a0-4841-adc4-69bfd00f7978}
I160928 18:50:24.130255 54 storage/replica_raftstorage.go:447  generated snapshot d8a1496f for range 2 at index 14 in 185.301µs.
I160928 18:50:24.135490 10146 storage/raft_transport.go:422  raft transport stream to node 1 established
F160928 18:50:24.140751 376 storage/replica.go:2361  [s1,r2:{"m"-/Max}] programming error: snapshot message from Raft.Ready 96ecfeaf doesn't match outgoing snapshot UUID d8a1496f.
goroutine 376 [running]:
github.com/cockroachdb/cockroach/util/log.getStacks(0x101, 0x863b2f0, 0x2440100, 0xc421060b58)
        /home/tschottdorf/src/github.com/cockroachdb/cockroach/util/log/clog.go:880 +0xa5
github.com/cockroachdb/cockroach/util/log.(*loggingT).outputLogEntry(0x2441200, 0xc400000003, 0x1e17f53, 0x12, 0x939, 0xc4208fe700, 0x7e)
        /home/tschottdorf/src/github.com/cockroachdb/cockroach/util/log/clog.go:745 +0x786
github.com/cockroachdb/cockroach/util/log.addStructured(0x7fc267b91280, 0xc420a3acf0, 0x3, 0x2, 0x178a4b6, 0x5f, 0xc421060f00, 0x2, 0x2)
        /home/tschottdorf/src/github.com/cockroachdb/cockroach/util/log/structured.go:72 +0x2a7
github.com/cockroachdb/cockroach/util/log.logDepth(0x7fc267b91280, 0xc420a3acf0, 0x1, 0xc400000003, 0x178a4b6, 0x5f, 0xc421060f00, 0x2, 0x2)
        /home/tschottdorf/src/github.com/cockroachdb/cockroach/util/log/log.go:88 +0x8c
github.com/cockroachdb/cockroach/util/log.Fatalf(0x7fc267b91280, 0xc420a3acf0, 0x178a4b6, 0x5f, 0xc421060f00, 0x2, 0x2)
        /home/tschottdorf/src/github.com/cockroachdb/cockroach/util/log/log.go:172 +0x7e
github.com/cockroachdb/cockroach/storage.(*Replica).sendRaftMessage(0xc420202600, 0x7, 0x3, 0x1, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/replica.go:2361 +0x93e
github.com/cockroachdb/cockroach/storage.(*Replica).handleRaftReadyRaftMuLocked(0xc420202600, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/replica.go:1907 +0x680
github.com/cockroachdb/cockroach/storage.(*Replica).handleRaftReady(0xc420202600, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/replica.go:1772 +0xcd
github.com/cockroachdb/cockroach/storage.(*Store).processReady(0xc420580000, 0x2)
        /home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/store.go:2766 +0x13b
github.com/cockroachdb/cockroach/storage.(*raftScheduler).worker(0xc42017a0e0, 0xc4201b41b0)
        /home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/scheduler.go:193 +0x331
github.com/cockroachdb/cockroach/storage.(*raftScheduler).Start.func1()
        /home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/scheduler.go:160 +0x33
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc4201b41b0, 0xc4200f7ae0)
        /home/tschottdorf/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x7d
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
        /home/tschottdorf/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:189 +0x66
```

**Comment by jordanlewis** [comment]
This is very sketchy and suspicious.

**Comment by tschottdorf** [comment]
```
I161003 14:18:46.622147 1 rand.go:76  Random seed: 8520730863382036289
I161003 14:18:46.626155 21 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I161003 14:18:46.686688 21 gossip/gossip.go:234  initial resolvers: []
W161003 14:18:46.686721 21 gossip/gossip.go:1045  no resolvers found; use --join to specify a connected node
I161003 14:18:46.686805 21 gossip/gossip.go:268  NodeID set to 1
I161003 14:18:46.689271 21 gossip/gossip.go:268  NodeID set to 1
I161003 14:18:46.690337 21 storage/store.go:1018  [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I161003 14:18:46.690434 21 gossip/gossip.go:282  NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:39220" > attrs:<>
I161003 14:18:46.695401 82 storage/replica_trigger.go:254  [s1,r1:/M{in-ax}] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 900ms following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=1970-01-01 00:00:00 +0000 UTC]
I161003 14:18:46.696073 21 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I161003 14:18:46.696924 21 gossip/gossip.go:234  initial resolvers: [127.0.0.1:39220]
W161003 14:18:46.696951 21 gossip/gossip.go:1047  no incoming or outgoing connections
I161003 14:18:46.697020 21 gossip/gossip.go:268  NodeID set to 2
I161003 14:18:46.710917 21 gossip/gossip.go:268  NodeID set to 2
I161003 14:18:46.715424 242 gossip/client.go:125  node 2: started gossip client to 127.0.0.1:39220
I161003 14:18:46.715652 21 storage/store.go:1018  [n2,s2]: failed initial metrics computation: [n2,s2]: system config not yet available
I161003 14:18:46.715716 21 gossip/gossip.go:282  NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:42638" > attrs:<>
I161003 14:18:46.718583 21 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I161003 14:18:46.719554 21 gossip/gossip.go:234  initial resolvers: [127.0.0.1:39220]
W161003 14:18:46.719580 21 gossip/gossip.go:1047  no incoming or outgoing connections
I161003 14:18:46.719631 21 gossip/gossip.go:268  NodeID set to 3
I161003 14:18:46.721230 21 gossip/gossip.go:268  NodeID set to 3
I161003 14:18:46.721499 21 storage/store.go:1018  [n3,s3]: failed initial metrics computation: [n3,s3]: system config not yet available
I161003 14:18:46.721566 21 gossip/gossip.go:282  NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:39868" > attrs:<>
I161003 14:18:46.722339 21 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I161003 14:18:46.722973 254 gossip/client.go:125  node 3: started gossip client to 127.0.0.1:39220
I161003 14:18:46.723325 21 gossip/gossip.go:234  initial resolvers: [127.0.0.1:39220]
W161003 14:18:46.723352 21 gossip/gossip.go:1047  no incoming or outgoing connections
I161003 14:18:46.723391 21 gossip/gossip.go:268  NodeID set to 4
I161003 14:18:46.729965 21 gossip/gossip.go:268  NodeID set to 4
I161003 14:18:46.730296 21 storage/store.go:1018  [n4,s4]: failed initial metrics computation: [n4,s4]: system config not yet available
I161003 14:18:46.730365 21 gossip/gossip.go:282  NodeDescriptor set to node_id:4 address:<network_field:"tcp" address_field:"127.0.0.1:33541" > attrs:<>
I161003 14:18:46.731025 202 gossip/client.go:125  node 4: started gossip client to 127.0.0.1:39220
I161003 14:18:46.733592 21 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I161003 14:18:46.734744 21 gossip/gossip.go:234  initial resolvers: [127.0.0.1:39220]
W161003 14:18:46.734764 21 gossip/gossip.go:1047  no incoming or outgoing connections
I161003 14:18:46.734814 21 gossip/gossip.go:268  NodeID set to 5
I161003 14:18:46.735829 21 gossip/gossip.go:268  NodeID set to 5
I161003 14:18:46.736015 21 storage/store.go:1018  [n5,s5]: failed initial metrics computation: [n5,s5]: system config not yet available
I161003 14:18:46.736058 21 gossip/gossip.go:282  NodeDescriptor set to node_id:5 address:<network_field:"tcp" address_field:"127.0.0.1:45672" > attrs:<>
I161003 14:18:46.736807 660 gossip/client.go:125  node 5: started gossip client to 127.0.0.1:39220
I161003 14:18:46.737402 187 gossip/server.go:252  node 1: refusing gossip from node 5 (max 3 conns); forwarding to 4 ({tcp 127.0.0.1:33541})
I161003 14:18:46.737569 660 gossip/client.go:130  node 5: closing client to node 1 (127.0.0.1:39220): received forward from node 1 to 4 (127.0.0.1:33541)
I161003 14:18:46.737765 476 gossip/client.go:125  node 5: started gossip client to 127.0.0.1:33541
I161003 14:18:46.742169 21 storage/replica_raftstorage.go:447  [s1,r1:/M{in-ax}] generated snapshot 5b1e7611 for range 1 at index 12 in 75.52µs.
I161003 14:18:46.743307 21 storage/store.go:2712  streamed snapshot: kv pairs: 27, log entries: 2
I161003 14:18:46.743690 699 storage/replica_raftstorage.go:577  [s2] [n2,s2,r1:{-}]: with replicaID [?], applying preemptive snapshot at index 12 (id=5b1e7611, encoded size=16, 1 rocksdb batches, 2 log entries)
I161003 14:18:46.745020 699 storage/replica_raftstorage.go:580  [s2] [n2,s2,r1:/M{in-ax}]: with replicaID [?], applied preemptive snapshot in 0.001s
I161003 14:18:46.745578 21 storage/replica_command.go:3278  change replicas: read existing descriptor range_id:1 start_key:"" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2
I161003 14:18:46.746727 703 storage/replica.go:1701  [s1,r1:/M{in-ax}] proposing ADD_REPLICA {NodeID:2 StoreID:2 ReplicaID:2} for range 1: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2}]
I161003 14:18:46.748167 21 storage/replica_raftstorage.go:447  [s1,r1:/M{in-ax}] generated snapshot 0f427129 for range 1 at index 14 in 57.391µs.
I161003 14:18:46.749129 21 storage/store.go:2712  streamed snapshot: kv pairs: 30, log entries: 4
I161003 14:18:46.749454 666 storage/replica_raftstorage.go:577  [s3] [n3,s3,r1:{-}]: with replicaID [?], applying preemptive snapshot at index 14 (id=0f427129, encoded size=16, 1 rocksdb batches, 4 log entries)
I161003 14:18:46.750437 666 storage/replica_raftstorage.go:580  [s3] [n3,s3,r1:/M{in-ax}]: with replicaID [?], applied preemptive snapshot in 0.001s
I161003 14:18:46.752701 21 storage/replica_command.go:3278  change replicas: read existing descriptor range_id:1 start_key:"" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:2 store_id:2 replica_id:2 > next_replica_id:3
I161003 14:18:46.759533 749 storage/raft_transport.go:422  raft transport stream to node 1 established
I161003 14:18:46.767665 460 storage/replica_trigger.go:438  [s2,r1:/M{in-ax}] unable to gossip first range; hasLease=false, err=<nil>
I161003 14:18:46.768243 569 storage/replica.go:1701  [s1,r1:/M{in-ax}] proposing ADD_REPLICA {NodeID:3 StoreID:3 ReplicaID:3} for range 1: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2} {NodeID:3 StoreID:3 ReplicaID:3}]
I161003 14:18:46.771578 760 storage/replica_trigger.go:438  [s2,r1:/M{in-ax}] unable to gossip first range; hasLease=false, err=<nil>
I161003 14:18:46.772031 763 storage/raft_transport.go:422  raft transport stream to node 1 established
I161003 14:18:46.772810 21 storage/replica_command.go:2364  initiating a split of this range at key "m" [r2]
I161003 14:18:46.773396 768 storage/replica_trigger.go:438  [s3,r1:/M{in-ax}] unable to gossip first range; hasLease=false, err=<nil>
I161003 14:18:46.777653 21 storage/replica_raftstorage.go:447  [s1,r2:{"m"-/Max}] generated snapshot dd404168 for range 2 at index 10 in 72.064µs.
I161003 14:18:46.778137 21 storage/store.go:2712  streamed snapshot: kv pairs: 24, log entries: 0
I161003 14:18:46.778351 771 storage/replica_raftstorage.go:577  [s4] [n4,s4,r2:{-}]: with replicaID [?], applying preemptive snapshot at index 10 (id=dd404168, encoded size=16, 1 rocksdb batches, 0 log entries)
I161003 14:18:46.782768 771 storage/replica_raftstorage.go:580  [s4] [n4,s4,r2:{"m"-/Max}]: with replicaID [?], applied preemptive snapshot in 0.004s
I161003 14:18:46.783290 21 storage/replica_command.go:3278  change replicas: read existing descriptor range_id:2 start_key:"m" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:2 store_id:2 replica_id:2 > replicas:<node_id:3 store_id:3 replica_id:3 > next_replica_id:4
W161003 14:18:46.787372 786 storage/stores.go:214  range not contained in one range: [/Meta2/Max,"m\x00"), but have [/Min,"m")
I161003 14:18:46.789223 359 storage/replica.go:1701  [s1,r2:{"m"-/Max}] proposing ADD_REPLICA {NodeID:4 StoreID:4 ReplicaID:4} for range 2: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2} {NodeID:3 StoreID:3 ReplicaID:3} {NodeID:4 StoreID:4 ReplicaID:4}]
I161003 14:18:46.799865 360 storage/replica_raftstorage.go:447  [s1,r2:{"m"-/Max}] generated snapshot 7c54e4eb for range 2 at index 14 in 59.902µs.
W161003 14:18:46.799942 129 storage/replica.go:1933  [s1,r2:{"m"-/Max}] Abandoning unsent snapshot {7c54e4eb-6b0b-4a49-ae0d-676ab8c108aa}
F161003 14:18:46.800054 129 storage/replica.go:2372  [s1,r2:{"m"-/Max}] programming error: snapshot message from Raft.Ready 7c54e4eb doesn't match outgoing snapshot UUID 00000000.
goroutine 129 [running]:
github.com/cockroachdb/cockroach/util/log.getStacks(0x0, 0x2fafdc2f, 0x245fd20, 0xc420b78b48)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/util/log/clog.go:880 +0xa5
github.com/cockroachdb/cockroach/util/log.(*loggingT).outputLogEntry(0x2460e20, 0xc400000003, 0x1e36f43, 0x12, 0x944, 0xc4210b8800, 0x7e)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/util/log/clog.go:743 +0x7aa
github.com/cockroachdb/cockroach/util/log.addStructured(0x2b6772952500, 0xc420791e00, 0x3, 0x2, 0x179ec7c, 0x5f, 0xc420b78f00, 0x2, 0x2)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/util/log/structured.go:72 +0x2a7
github.com/cockroachdb/cockroach/util/log.logDepth(0x2b6772952500, 0xc420791e00, 0x1, 0xc400000003, 0x179ec7c, 0x5f, 0xc420b78f00, 0x2, 0x2)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/util/log/log.go:96 +0x8c
github.com/cockroachdb/cockroach/util/log.Fatalf(0x2b6772952500, 0xc420791e00, 0x179ec7c, 0x5f, 0xc420b78f00, 0x2, 0x2)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/util/log/log.go:180 +0x7e
github.com/cockroachdb/cockroach/storage.(*Replica).sendRaftMessage(0xc4201a1180, 0x7, 0x2, 0x1, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/replica.go:2372 +0x971
github.com/cockroachdb/cockroach/storage.(*Replica).handleRaftReadyRaftMuLocked(0xc4201a1180, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/replica.go:1918 +0x683
github.com/cockroachdb/cockroach/storage.(*Replica).handleRaftReady(0xc4201a1180, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/replica.go:1783 +0xcd
github.com/cockroachdb/cockroach/storage.(*Store).processReady(0xc4201b8840, 0x2)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/store.go:2771 +0x13b
github.com/cockroachdb/cockroach/storage.(*raftScheduler).worker(0xc420199340, 0xc4201ae1b0)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/scheduler.go:193 +0x331
github.com/cockroachdb/cockroach/storage.(*raftScheduler).Start.func1()
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/scheduler.go:160 +0x33
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc4201ae1b0, 0xc42028d5a0)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x7d
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:189 +0x66


ERROR: exit status 255

9106 runs completed, 1 failures, over 1m58s
FAIL
Makefile:146: recipe for target 'stress' failed
make: *** [stress] Error 1
```

**Comment by tschottdorf** [comment]
@jordanlewis can you get to the bottom of this?

**Comment by tamird** [comment]
Adjacent lines in the log:
```
W161003 14:18:46.799942 129 storage/replica.go:1933  [s1,r2:{"m"-/Max}] Abandoning unsent snapshot {7c54e4eb-6b0b-4a49-ae0d-676ab8c108aa}
F161003 14:18:46.800054 129 storage/replica.go:2372  [s1,r2:{"m"-/Max}] programming error: snapshot message from Raft.Ready 7c54e4eb doesn't match outgoing snapshot UUID 00000000.
```

Seems promising.

**Comment by jordanlewis** [comment]
Looking at this no.

**Comment by jordanlewis** [comment]
@tschottdorf was this latest failure also with the `deadlock` tag? It shouldn't matter, but I'm curious about the reproduction steps.

**Comment by tschottdorf** [comment]
No, that was just `make stress PKG=./storage TESTS=TestStoreRangeDownReplicate` on my gceworker.

**Comment by jordanlewis** [comment]
I haven't managed to track this down yet. I think it has something to do with preemptive snapshots racing with non-preemptive snapshots but I haven't proven that - I'll pick this up in a week when I'm back from vacation. If anyone else would like to grab this in the meantime, please feel free.

**Comment by tamird** [comment]
Looks like we've been saving it for you.

**Comment by tamird** [comment]
I believe this was actually fixed by #10875.
